### PR TITLE
Replace `len()` Zero Checks with Pythonic Empty Sequence Checks

### PR DESCRIPTION
**Pull request by andylizf**
Replace len() comparisons with 0/1 with pythonic empty sequence checks, following [Google Python Style Guide's recommendation](https://google.github.io/styleguide/pyguide.html#2144-decision):

> For sequences (strings, lists, tuples), use the fact that empty sequences are false, so `if seq:` and `if not seq:` are preferable to `if len(seq):` and `if not len(seq):` respectively.

Also fixed some typing annotations along the way.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

**Comment by andylizf** [comment]
@cblmemo PTAL, thanks!

**Comment by Michaelvll** [comment]
Please refer to the style guide: https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations

**Comment by andylizf** [comment]
@Michaelvll By the way, when you have a moment, could you take a look at #4250 ? Thanks a lot!

**Comment by Michaelvll** [comment]
Ahh, thanks for fixing these! I misread the message. Looks pretty good at the first glance. cc'ing @cblmemo to take a closer look. : )

File: format.sh [hunk 1]
```diff
@@ -132,7 +132,7 @@ elif [[ "$1" == '--all' ]]; then
     pylint "${PYLINT_FLAGS[@]}" sky
 else
     # Pylint only files in sky/ that have changed in last commit.
-    changed_files=$(git diff --name-only --diff-filter=ACM "$MERGEBASE" -- 'sky/*.py' 'sky/*.pyi')
+    changed_files=$(git diff --name-only --diff-filter=ACM "$MERGEBASE" -- 'sky/*.py' 'sky/*.pyi' ':!sky/skylet/**')
```
Comment (cblmemo) [inline format.sh:135]: what is this for

File: sky/clouds/service_catalog/common.py [hunk 2]
```diff
@@ -680,7 +680,7 @@ def get_image_id_from_tag_impl(df: 'pd.DataFrame', tag: str,
     df = _filter_region_zone(df, region, zone=None)
     assert len(df) <= 1, ('Multiple images found for tag '
                           f'{tag} in region {region}')
-    if len(df) == 0:
+    if not df:
```
Comment (cblmemo) [inline sky/clouds/service_catalog/common.py:683]: Not sure if we could do this to dfs. Pls also check other places in this PR

```bash
>>> df = pd.DataFrame()
>>> bool(df)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/txia/miniconda3/envs/sky-serve/lib/python3.9/site-packages/pandas/core/generic.py", line 1466, in __nonzero__
    raise ValueError(
ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

File: sky/clouds/service_catalog/common.py [hunk 3]
```diff
@@ -694,4 +694,4 @@ def is_image_tag_valid_impl(df: 'pd.DataFrame', tag: str,
     df = df[df['Tag'] == tag]
     df = _filter_region_zone(df, region, zone=None)
     df = df.dropna(subset=['ImageId'])
-    return len(df) > 0
+    return bool(df)
```
Comment (cblmemo) [inline sky/clouds/service_catalog/common.py:697]: same as above

File: sky/resources.py [hunk 4]
```diff
@@ -1008,7 +1008,7 @@ def _try_validate_labels(self) -> None:
                         f'Label rejected due to {cloud}: {err_msg}'
                     ])
                     break
-        if len(invalid_table.rows) > 0:
+        if invalid_table.rows:
```
Comment (cblmemo) [inline sky/resources.py:1011]: same, lets check if the `rows` attribute can be converted this way.

File: sky/serve/serve_utils.py [hunk 5]
```diff
@@ -785,8 +785,6 @@ def get_endpoint(service_record: Dict[str, Any]) -> str:
     handle = global_user_state.get_handle_from_cluster_name(
         SKY_SERVE_CONTROLLER_NAME)
     assert isinstance(handle, backends.CloudVmRayResourceHandle)
-    if handle is None:
-        return '-'
```
Comment (cblmemo) [inline sky/serve/serve_utils.py:789]: Why remove this?

File: sky/serve/serve_utils.py [hunk 6]
```diff
@@ -885,13 +883,12 @@ def _format_replica_table(replica_records: List[Dict[str, Any]],
         zone = '-'
 
         replica_handle: 'backends.CloudVmRayResourceHandle' = record['handle']
-        if replica_handle is not None:
```
Comment (cblmemo) [inline sky/serve/serve_utils.py:888]: Why remove this?

**Review by cblmemo — COMMENTED** [review]
Thanks for refactoring this @andylizf ! Mostly looks good to me. Left some discussions.

Also, pls run smoke test to make sure it does not break anything ;) (reminder: Pls run the smoke tests on GCP. to do so, you can temporarily disable aws on your laptop.

File: sky/resources.py (hunk 4 above)
Comment (andylizf) [reply sky/resources.py:1011]: `rows` is a `list`, and the answer is yes

File: format.sh (hunk 1 above)
Comment (andylizf) [reply format.sh:135]: found some files in `skylet` are not properly formatted.

File: sky/serve/serve_utils.py (hunk 5 above)
Comment (andylizf) [reply sky/serve/serve_utils.py:789]: handle is always non-None

File: format.sh (hunk 1 above)
Comment (cblmemo) [reply format.sh:135]: Pls double check.

File: sky/serve/serve_utils.py (hunk 5 above)
Comment (cblmemo) [reply sky/serve/serve_utils.py:789]: move assert after this if branch

File: sky/serve/serve_utils.py (hunk 6 above)
Comment (cblmemo) [reply sky/serve/serve_utils.py:888]: Add optional in the type annotation

File: sky/clouds/service_catalog/common.py (hunk 2 above)
Comment (andylizf) [reply sky/clouds/service_catalog/common.py]: Checked by
```bash
andyl@DESKTOP-7FP6SMO ~/skypilot (fix-serve-types)> git diff upstream/master --name-only | xargs grep -l "DataFrame"                                            (skypilot) 
sky/cli.py
sky/clouds/service_catalog/common.py
sky/clouds/service_catalog/data_fetchers/fetch_azure.py
sky/clouds/service_catalog/data_fetchers/fetch_vsphere.py
```

**Comment by cblmemo** [comment]
@andylizf , could you help resolve the merge conflicts?

File: sky/serve/serve_utils.py [hunk 7]
```diff
@@ -383,7 +383,7 @@ def _get_service_status(
 
 
 def get_service_status_encoded(service_names: Optional[List[str]]) -> str:
-    service_statuses = []
+    service_statuses: List[Dict[str, Any]] = []
```
Comment (cblmemo) [inline sky/serve/serve_utils.py:386]:
```suggestion
    service_statuses: List[Dict[str, str]] = []
```

**Review by cblmemo — APPROVED** [review]
Thanks! LGTM. Left one nit

**Pull request merged**